pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -611,13 +613,63 @@ func (m *kubeGenericRuntimeManager) killContainer(pod *v1.Pod, containerID kubec
 	return err
 }
 
+func isSidecar(containers []v1.Container, name string) bool {
+	for _, container := range containers {
+		if container.Name == name && lifecycleSidecar(container) {
+			return true
+		}
+	}
+	return false
+}
+
+func gracePeriodRemaining(pod *v1.Pod, gracePeriodOverride *int64, startTime time.Time) *int64 {
+	timeElapsed := int64(time.Since(startTime).Seconds())
+
+	switch {
+	case gracePeriodOverride != nil:
+		newPeriod := *gracePeriodOverride - timeElapsed
+		gracePeriodOverride = &newPeriod
+	case pod.DeletionGracePeriodSeconds != nil:
+		newPeriod := *pod.DeletionGracePeriodSeconds - timeElapsed
+		gracePeriodOverride = &newPeriod
+	case pod.Spec.TerminationGracePeriodSeconds != nil:
+		newPeriod := *pod.Spec.TerminationGracePeriodSeconds - timeElapsed
+		gracePeriodOverride = &newPeriod
+	}
+
+	return gracePeriodOverride
+}
+
 // killContainersWithSyncResult kills all pod's containers with sync results.
 func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, runningPod kubecontainer.Pod, gracePeriodOverride *int64) (syncResults []*kubecontainer.SyncResult) {
 	containerResults := make(chan *kubecontainer.SyncResult, len(runningPod.Containers))
 	wg := sync.WaitGroup{}
 
-	wg.Add(len(runningPod.Containers))
-	for _, container := range runningPod.Containers {
+	containers := runningPod.Containers
+	var sidecars []*kubecontainer.Container
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) {
+		if pod != nil {
+			var nonSidecars []*kubecontainer.Container
+			for _, container := range runningPod.Containers {
+				if isSidecar(pod.Spec.Containers, container.Name) {
+					sidecars = append(sidecars, container)
+				} else {
+					nonSidecars = append(nonSidecars, container)
+				}
+			}
+			containers = nonSidecars
+		}
+		if len(sidecars) != 0 {
+			timePreStopStarted := time.Now()
+			m.preStopContainers(pod, sidecars)
```

Comment:
One thing I'm noticing is that `m.preStopContainers` has a decent amount of duplication with the code in `killContainer`. I know that `killContainer` is responsible for both running the pre stop and killing the container, and Sidecar containers want to perform those as separate entities. However, I wonder if there's an opportunity to extract shared subfunctions from `killContainer`? Maybe `runPreStop` and `killActualContainer` (bad name)? And then the we wouldn't need to essentially duplicate this functionality for sidecars?

Also, possible I'm missing a difference in the implementation :)

I _think_

Reply:
Yeah there is quite a bit of overlap, the restoring/fetching pod/container spec and executing the prestop are both shared between the functions. I guess I was somewhat cautious about changing a function like `killContainer` if I didn't have to, as you can't really feature gate something like that and I wouldn't want to inadvertently break a rather important thing with my alpha feature. 
I can have a go at extracting some of the common functionality and see what it looks like though.

Reply:
I think that's very well thought through @Joseph-Irving :)

Perhaps would I'd recommend is the following - maybe give a shot to extracting common functionality and see if it can be done in a way that we are confident with and would still be useful even if we ended up deleting all sidecar code after the alpha. If the answer is yes, great. If the answer is no, I agree we don't want to disrupt core functionality for an alpha feature. Perhaps in that case, we could just add a refactor as part of the graduation criteria for GA?

Reply:
It's a waste of time that run sidecar's preStop hook twice